x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+            for (SourceDestValidation validation : START_AND_PUT_VALIDATIONS) {
+                if (shouldDefer && validation.isDeferrable()) {
+                    continue;
+                }
+                validation.validate(context);
+            }
+
+        } catch (ValidationException e) {}
+
+        return context.getValidationException();
+    }
+
+    static class SourceMissingValidation implements SourceDestValidation {
+
+        @Override
+        public boolean isDeferrable() {
+            return true;
+        }
+
+        @Override
+        public void validate(Context context) {
+            // boolean remoteIndexesEmpty = context.isRemoteSearchEnabled() ? context.resolveRemoteSource().isEmpty() : true;
+
+            try {
+                // non-trivia: if source contains a wildcard index, which does not resolve to a concrete index
+                // the resolved indices might be empty, but we can check if source contained something, this works because
+                // of no wildcard index is involved the resolve would have thrown an exception
+                if (context.resolveSource().isEmpty() && context.resolveRemoteSource().isEmpty() && context.getSource().length == 0) {
+                    context.addValidationError(SOURCE_INDEX_MISSING, true, Strings.arrayToCommaDelimitedString(context.getSource()));
+                }
+            } catch (IndexNotFoundException e) {
+                context.addValidationError(e.getMessage(), true);
+            }
+        }
+    }
+
+    static class RemoteSourceEnabledAndRemoteLicenseValidation implements SourceDestValidation {
+        @Override
+        public boolean isDeferrable() {
+            return true;
+        }
+
+        @Override
+        public void validate(Context context) {
+            if (context.resolveRemoteSource().isEmpty()) {
+                return;
+            }
+
+            List<String> remoteIndices = new ArrayList<>(context.resolveRemoteSource());
+            // we can only check this node at the moment, clusters with mixed CCS enabled/disabled nodes are not supported,
+            // see gh#50033
+            if (context.isRemoteSearchEnabled() == false) {
+                context.addValidationError(NEEDS_REMOTE_CLUSTER_SEARCH, true, context.resolveRemoteSource(), context.getNodeName());
+                return;
+            }
+
+            // this can throw
+            List<String> remoteAliases;
+            try {
+                remoteAliases = RemoteClusterLicenseChecker.remoteClusterAliases(context.getRegisteredRemoteClusterNames(), remoteIndices);
+            } catch (NoSuchRemoteClusterException e) {
+                context.addValidationError(e.getMessage(), true);
+                return;
+            } catch (Exception e) {
+                context.addValidationError(ERROR_REMOTE_CLUSTER_SEARCH, true, e.getMessage());
+                return;
+            }
+
+            CountDownLatch latch = new CountDownLatch(1);
```

Comment:
We shouldn't do this. This could be executing on a network thread (i.e. from a transport class).

I think a listener should be passed into the class. Possibly only if remote cluster checks are necessary? Or probably always for simplicity's sake.